### ISSUE TITLE
Parsing custom message headers

### ISSUE THREAD

**Issue by anatol** [lestrrat-go/jwx#205]
Hi, I am trying  to deal with `clevis` encryption in go. I have a following json message:

```json
{"alg":"ECDH-ES","clevis":{"pin":"tang","tang":{"adv":{"keys":[{"alg":"ES512","crv":"P-521","key_ops":["verify"],"kty":"EC","x":"AGeZ7xh05hGfO8HEaDYCr855kq8FEw5Hxzo1Ne30A7frqQq0u_hbn_Vo9lUaLAwce8o2a-soEKGa7pwmIeFUm7Io","y":"AEUeuTA_Ub4u605OGkDrpQHFCqXhIjPhTZAIVFIrdpU-iEQ_7rzyBwH2OQJMvScmkRk5UtfHxnUskbHU7DIzX89C"},{"alg":"ECMR","crv":"P-521","key_ops":["deriveKey"],"kty":"EC","x":"AJwCS845x9VljR-fcrN2WMzIJHDYuLmFShhyu8ci14rmi2DMFp8txIvaxG8n7ZcODeKIs1EO4E_Bldm_pxxs8cUn","y":"ASjz754cIQHPJObihPV8D7vVNfjp_nuwP76PtbLwUkqTk9J1mzCDKM3VADEk-Z1tP-DHiwib6If8jxnb_FjNkiLJ"}]},"url":"http://localhost"}},"enc":"A256GCM","epk":{"crv":"P-521","kty":"EC","x":"AQwaFXAUbmSlmh65-921zAKP5t7jpPgH44Gxu4CNd1DEL6vjJm5eeVr4DOFzQKFr7GB7LsklYIjcqmRBn8uTaQS5","y":"AbtYf1Kk9aacBjVn7mah2o5hqp1lGTRQUuMxW4qgjlygNRm4-x8WdWHOJbMv2Xa574dtpW3LOnzcLjyWsG2ASXEb"},"kid":"2Mc_43O_BOrOJTNrGX7uJ6JsIYE"}
```

There is a custom header node called `clevis` that contains required information such as keys plus some custom fields I need access to.

Right now I am trying to parse this message above with
```go
msg, err := jwe.ParseString(data)
if err != nil {
	return nil, err
}
```
but `msg` does not contain any information about these extra headers.

In other implementations I can do `msg.Header.ExtraHeaders["clevis"].(map[string]interface{})` but I am not sure what is the recommended way to deal with it in `jwx` library.

**Comment by lestrrat** [comment]
Since you're using `jwe`, did you check `PrivateParams()`?

**Comment by anatol** [comment]
Thank you for your response @lestrrat.

Could you please provide an example of the function usage? Sorry, I am really new to this library and can't figure out what object supports `PrivateParams()`.

I grepped `jwx` project and I see that the method you mentioned is defined on `stdHeaders/ecdsaPrivateKey/ecdsaPublicKey/...` and other private structures that are not exposed outside of `jwx` project namespace.

**Comment by lestrrat** [comment]
https://github.com/lestrrat-go/jwx/blob/master/jwe/headers_gen.go#L273

`msg.Headers().PrivateParams()` would  return a `map[string]interface{}`, and the rest is the same from your earlier example :)

**Comment by anatol** [comment]
Hi, `msg` is of type `jwe.Message` that does not have a public `Headers()` method. It has `UnprotectedHeaders()` method though that returns `jwe.Headers`. But that interface is defined as

```go
type Headers interface {
	AgreementPartyUInfo() buffer.Buffer
	AgreementPartyVInfo() buffer.Buffer
	Algorithm() jwa.KeyEncryptionAlgorithm
	Compression() jwa.CompressionAlgorithm
	ContentEncryption() jwa.ContentEncryptionAlgorithm
	ContentType() string
	Critical() []string
	EphemeralPublicKey() jwk.ECDSAPublicKey
	JWK() jwk.Key
	JWKSetURL() string
	KeyID() string
	Type() string
	X509CertChain() []string
	X509CertThumbprint() string
	X509CertThumbprintS256() string
	X509URL() string
	Iterate(ctx context.Context) Iterator
	Walk(ctx context.Context, v Visitor) error
	AsMap(ctx context.Context) (map[string]interface{}, error)
	Get(string) (interface{}, bool)
	Set(string, interface{}) error
	Remove(string) error
	Encode() ([]byte, error)
	Decode([]byte) error
}
```

and it does not have any `PrivateParams()`.

The code you pointed is

```go
func (h *stdHeaders) PrivateParams() map[string]interface{} {
	return h.privateParams
}
```

and it defines method on `stdHeaders` that is **private** structure.

**Comment by lestrrat** [comment]
I meant ProtectedHeaders https://github.com/lestrrat-go/jwx/blob/master/jwe/message.go#L123

and if type Headers doesn't have `PrivateParams`, that's a bug :/

**Comment by lestrrat** [comment]
Uh-huh...

```
lestrrat@finch jwe % go test -v .
# github.com/lestrrat-go/jwx/jwe_test [github.com/lestrrat-go/jwx/jwe.test]
./headers_test.go:83:10: h.PrivateParams undefined (type jwe.Headers has no field or method PrivateParams)
FAIL	github.com/lestrrat-go/jwx/jwe [build failed]
FAIL
```

**Comment by anatol** [comment]
I use a debugger to inspect the datastructures and it actually `msg.Recipients()[0].Headers()` that contains params I need. But this param is not exposed via a public interface.

Is there a chance you can add `PrivateParams()` to `jwe.Headers` type?

**Comment by lestrrat** [comment]
Recipients! IIRCl that part (which parts those headers belong to) actually depends on the original message, so okay, I was wrong.

As far as exposing `PrivateParams()`, can you please check #206 ?

**Comment by anatol** [comment]
@lestrrat that branch works for me, thank you!

**Comment by lestrrat** [comment]
Cool. Sorry about `PrivateParams` not being exposed. `PrivateParams` has existed forever, but a few months ago we did a whole rewrite and I guess that just went unnoticed. I'm going to double check if it works for other packages and then merge to master

**Comment by anatol** [comment]
No worries. And thank you very much for the quick response!

**Comment by lestrrat** [comment]
BTW, now that I think about it, you should be able to also access the key using `AsMap()` method, which returns a `map[string]interface{}` with the entire key set. But yeah, either way PrivateParams should have existed.

**Comment by lestrrat** [comment]
fixed in #206
